crates/crypto/bls/src/supranational/signature.rs

```
@@ -0,0 +1,145 @@
+            .map_err(|e| anyhow::anyhow!("Failed to convert signature: {:?}", e))
+    }
+
+    /// Verifies a BLS signature against a public key and message.
+    ///
+    /// This function will return `false` in any error case, including:
+    /// - If the signature bytes are invalid or malformed
+    /// - If the public key bytes are invalid or malformed
+    /// - If the actual signature verification fails
+    ///
+    /// # Arguments
+    /// * `pubkey` - The public key to verify against
+    /// * `message` - The message that was signed
+    ///
+    /// # Returns
+    /// `bool` - true if the signature is valid, false otherwise
+    pub fn verify(&self, pubkey: &PubKey, message: &[u8]) -> bool {
+        self.to_blst_signature()
+            .and_then(|sig| pubkey.to_blst_pubkey().map(|pk| (sig, pk)))
+            .is_ok_and(|(sig, pk)| {
+                sig.verify(true, message, DST, &[], &pk, false) == blst::BLST_ERROR::BLST_SUCCESS
+            })
+    }
+
+    /// Verifies the signature against an aggregate of multiple public keys and a message
+    ///
+    /// # Arguments
+    /// * `pubkeys` - Collection of public key references to be aggregated
+    /// * `message` - Message that was signed
+    ///
+    /// # Returns
+    /// * `bool` - Returns true if the signature is valid for the aggregated public keys and
+    ///   message, false if the signature is invalid or if any error occurs during verification
+    pub fn fast_aggregate_verify<'a, P>(&self, pubkeys: P, message: &[u8]) -> bool
+    where
+        P: AsRef<[&'a PubKey]>,
+    {
+        self.to_blst_signature()
+            .and_then(|sig| {
+                let public_keys = pubkeys
+                    .as_ref()
+                    .iter()
+                    .map(|key| key.to_blst_pubkey())
+                    .collect::<Result<Vec<_>, _>>()?;
+                Ok((sig, public_keys))
+            })
+            .is_ok_and(|(sig, public_keys)| {
```

We shouldn't hide the error using an `is_ok_and` we should return it

---

Yes, I would like to discuss with this also. It's quite similar as we discussed at #103.

I think we should hide error in this case. All of the errors here are `BLST_ERROR`s, which is fully dependent on `blst`. This would not be available on other libraries like `zkcrypto/bls12_381` or else. Here is the `BLST_ERROR`.

```rust
#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
pub enum BLST_ERROR {
    BLST_SUCCESS = 0,
    BLST_BAD_ENCODING = 1,
    BLST_POINT_NOT_ON_CURVE = 2,
    BLST_POINT_NOT_IN_GROUP = 3,
    BLST_AGGR_TYPE_MISMATCH = 4,
    BLST_VERIFY_FAIL = 5,
    BLST_PK_IS_INFINITY = 6,
    BLST_BAD_SCALAR = 7,
}
```

Other reason for this is that returning `bool` type is more natural imho. The function signatures for `bls` module is provided like [this](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#bls-signatures). Developers don't have to think of what's going underneath this module.
Returning bool type is also useful to deal with the functions like `apply_deposit`. [`apply_deposit`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#deposits) doesn't throw an error if anything gone wrong. [`is_valid_indexed_attestation`](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#is_valid_indexed_attestation) also returns bool type. We could check by `ensure!` whether the verification result is true or not at the high level.

---

We can define our own error enum using this error https://docs.rs/thiserror/latest/thiserror/

> [apply_deposit](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#deposits) doesn't throw an error if anything gone wrong

That is fine we can handle the error in apply_deposit and choose to return ok if an error occurs, but we shouldn't hide the error, we should let the developer calling the functions on how to handle it, instead of choosing how to handle it for them

---

Understood. Then I'm gonna think of which errors should be presented, and then handle those errors at the caller frame.

---

https://github.com/ReamLabs/ream/pull/117/commits/84c9762a16350d5dd3222641841855bd25e8549b added a custom error.